### ISSUE TITLE
RFC: `next-urql` – send App props for pre-pass

### ISSUE THREAD

**Issue by tomvardasca** [urql-graphql/urql#2573]
## Summary

Using the `next-urql` library with `withUrqlClient` is used with `ssr: true`, all the Next App tree is re-rendered. When we use `withUrqlClient` in the `App` everything works fine. But when we use `withUrqlClient` in pages and have React Context providers in the custom `Aapp` that receive props from `App.getInitialProps`. In the pre-pass step, this values (it only fetches the `Page.getInitialProps`) and no App props from   `App.getInitialProps`. This can be an issue if, for example, we use a Feature Flag React context, the behavior in the pre-pass step will be different.

## Proposed Solution

What I want to propose is a callback inside `WithUrqlClientOptions` where we could send this app props from the Next context, something like:
```ts
export interface WithUrqlClientOptions {
  ssr?: boolean;
  neverSuspend?: boolean;
  staleWhileRevalidate?: boolean;
  getAppProps?: (ctx?: NextPageContext) => any;
}
```

And the only change would be in `with-urql-client.ts`:
```ts
       const appProps = isApp ? {} : options?.getAppProps?.(ctx) ?? {};
       // L: 163 
       const appTreeProps = isApp ? props : { pageProps: props, ...appProps };
```

With this, we are able to send the app props, and keep using the context in the pre-pass step.

**Comment by nruhe** [comment]
I believe we're running into a similar issue when trying to use withUrqlClient on _app.js alongside [appWithTranslations](appWithTranslation) from https://github.com/i18next/next-i18next

**Comment by Swiftwork** [comment]
@nruhe I opened an issue related to this but not quite the same regarding next-i18next. https://github.com/urql-graphql/urql/issues/2892

**Comment by kitten** [comment]
I have to be honest — I can't keep track of Next's convoluted way of defining their `getInitialProps`, so to keep it simple, is this obsolete with #2892 being resolved with #2909 having been merged?

**Comment by kitten** [comment]
Closing due to inactivity
